crates/oxc_linter_plugin/src/plugin.rs

```diff
@@ -0,0 +1,615 @@
+    query: NamedSource,
+    #[label = "This test failed."]
+    err_span: SourceSpan,
+    #[related]
+    errors: Vec<Report>,
+}
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("Test expected to fail, but passed.")]
+struct ExpectedTestToFailButPassed {
+    #[source_code]
+    query: NamedSource,
+    #[label = "This test should have failed but it passed."]
+    err_span: SourceSpan,
+}
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("Unexpected errors in fail test.")]
+struct UnexpectedErrorsInFailTest {
+    #[related]
+    errors: Vec<Report>,
+    #[source_code]
+    query: NamedSource,
+    #[label = "This test should have failed but it passed."]
+    err_span: SourceSpan,
+}
+
+enum PassOrFail {
+    Pass,
+    Fail,
+}
+
+fn span_of_test_n(
+    yaml_text: &str,
+    test_ix: usize,
+    test_code: &str,
+    pass_or_fail: &PassOrFail,
+) -> SourceSpan {
+    let yaml = YamlLoader::load_from_str(
+        // TODO: Should we just save the string after we read it originally?
+        yaml_text,
+    )
+    .expect("to be able to parse yaml for error reporting");
+    let YamlElt::Hash(hash) = &yaml.docs[0].yaml else {
+        unreachable!("must be a top level hashmap in the yaml")
+    };
+    let tests_hash_key = hash
+        .keys()
+        .find(|x| {
+            let YamlElt::String(str) = &x.yaml else { return false };
+            str == "tests"
+        })
+        .expect("to be able to find tests hash in yaml file");
+    let YamlElt::Hash(tests_hash) = &hash[tests_hash_key].yaml else {
+        unreachable!("there must be a tests hashmap in the yaml")
+    };
+    let pass_hash_key = tests_hash
+        .keys()
+        .find(|x| {
+            let YamlElt::String(str) = &x.yaml else { return false };
+            str == if matches!(pass_or_fail, PassOrFail::Pass) { "pass" } else { "fail" }
+        })
+        .expect("to be able to find pass hash in yaml file");
+    let YamlElt::Array(test_arr) = &tests_hash[pass_hash_key].yaml else {
+        unreachable!("there must be a pass array in the yaml")
+    };
+    let test_hash_span = &test_arr[test_ix].lines_range();
+    let start = yaml_text
+        .char_indices()
+        .filter(|a| a.1 == '\n')
+        .nth(test_hash_span.0 - 1) // subtract one because span is 1-based
+        .map(|a| a.0)
+        .expect("to find start of span of test");
+    let start_of_end = yaml_text[start..]
+        .find(&test_code[0..test_code.find('\n').unwrap_or(test_code.len())])
+        .expect("to find start of end")
+        + start;
+
+    let nl = test_code.chars().filter(|a| *a == '\n').count();
+    let end_of_end = yaml_text[start_of_end..]
+        .char_indices()
+        .filter(|a| a.1 == '\n')
+        .nth(nl - 1)
+        .map(|a| a.0)
+        .expect("to find end of end of span of test")
+        + start_of_end;
+
+    SourceSpan::new(start.into(), (end_of_end - start).into())
+}
+
+pub fn test_queries(queries_to_test: &PathBuf) -> oxc_diagnostics::Result<()> {
```

Comment:
The name here is confusing, and there are no docs to double-check: is this test code for this crate, or public API intended to be used externally?

Reply:
This function tests the queries at a path. I don’t think we want to expose it though as it’s probably best if end users just use something like insta on the output of running their custom lints on a directory. The testing situation for end users is tbd though as this is pr is not exposing this crate publicly, it’s just adding it to the workspace so it stays up to date with master.